src/main/scala/util/RecordMap.scala

```diff
@@ -0,0 +1,27 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.util
+
+import Chisel._
+import chisel3.Record
+import scala.collection.immutable.ListMap
+
+final case class RecordMap[T <: Data](eltMap: ListMap[String, T])
+    extends Record with collection.Seq[T] {
+
+  // Required by Seq
+  def apply(x: Int) = eltMap.values.toSeq(x)
+  def length = eltMap.size
+  def iterator = eltMap.values.iterator
+
+  def apply(x: String) = eltMap.get(x)
+
+  // Required by Record
+  val elements = eltMap
+
+  override def cloneType: this.type = (new RecordMap(eltMap.map{case (k, v) => k -> v.chiselCloneType})).asInstanceOf[this.type]
+
```

Comment:
```suggestion
  import chisel3.internal.requireIsChiselType
  import chisel3.experimental.internal.chiselTypeClone
  
  eltMap.foreach { case (name, elt) => requireIsChiselType(elt, name) }

  val elements = ListMap() ++ eltMap.mapValues(chiselTypeClone) // mapValues return value is lazy

  override def cloneType: this.type = (new RecordMap(eltMap)).asInstanceOf[this.type]

```

I know this is following the style of `HeterogeneousBag`, but I think this is wrong. In general, custom `Record` implementations should provide fresh types to `elements` rather than the ones passed by the user. This is because the `elements` must be _bound in place_ by Chisel, ie. they will be mutated. `HeterogeneousBag` did the "copy on clone" technique because it was intentionally allowing things to be used as hardware before the HeterogeneousBag was created. This was because all ports had to be packed into `val io`. Since that is no longer true, this pattern should no longer be used. Furthermore, the pattern won't work in `import chisel3._` anyway.

This suggestion also implies that we could make the `eltMap` argument an `Iterable[String, T]` and then provide a companion object with a varargs version as well.

Reply:
I tried the Iterable[String, T] but then that was annoying to write an accessor by the string name, which I assume we would want at some point. I still wrote the varargs version.

I did not use your suggestion directly, but I believe I captured the request... please take a look.

This doesnt work in my use case however, I get an exception on the chiselTypeClone call. WIll point it out on #2487